src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,398 @@
+    *                               reverse complementary sequence at the ends of reads.
+    * @param mean_length the mean length of the reverse complementary sequence in the `read_based_switchbacks`.
+    * @param mean_offset the mean offset of the `read_based_switchbacks`.
+    * @param tandem_based_switchbacks the count of switchback_templates that were identified because they had paired
+    *                                 reads mapped in FF or RR orientations within the designated gap size range.
+    * @param mean_gap the mean gap size for `tandem_based_switchbacks`
+    */
+  case class SwitchMetric(sample: String,
+                          library: String,
+                          templates: Count,
+                          aligned_templates: Count,
+                          switchback_templates: Count,
+                          fraction_switchbacks: Proportion,
+                          read_based_switchbacks: Count,
+                          mean_length: Double,
+                          mean_offset: Double,
+                          tandem_based_switchbacks: Count,
+                          mean_gap: Double
+                         ) extends Metric
+
+  /**
+    * Attempts to find a switchback event based on soft-clipped sequence within the primary reads of the template.
+    *
+    * @param refs a map of reference sequence name to [[ReferenceSequence]] object containing the sequence.
+    * @param template the template to be examined.
+    * @param minLength the minimum length of switchback to be detected.
+    * @param maxOffset the maximum offset on the reference between the end of the pre-switch sequence and the start
+    *                  of the post-switch sequence
+    * @param maxErrorRate the maximum rate of mismatches between the switched-back sequence and the reference
+    * @return Some(ReadBasedHit) if a hit is found else None.
+    */
+  private[bam] def findReadBasedSwitchback(refs: Map[String, ReferenceSequence],
+                                      template: Template,
+                                      minLength: Int,
+                                      maxOffset: Int,
+                                      maxErrorRate: Double): Option[ReadBasedHit] = {
+    if (maxOffset <= 0) None else template.primaryReads.filter(_.mapped).flatMap { rec =>
+        if (rec.positiveStrand && rec.cigar.head.operator == CigarOperator.S && rec.cigar.head.length >= minLength) {
+          val ref = refs(rec.refName)
+          val bases = Sequences.revcomp(rec.basesString.take(rec.cigar.head.length))
+          val windowStart = max(1, rec.start - maxOffset)
+          val windowEnd = min(ref.length(), rec.start + bases.length + maxOffset)
+          val refBases = new String(ref.getBases, windowStart - 1, windowEnd - windowStart + 1)
+          find(bases, refBases, maxErrorRate).map { hit => hit.read = Some(rec); hit.offset = -(hit.offset + windowStart - rec.start); hit }
```

Comment:
If you pass in `rec` to `find`, then `hit.read` can be a `val`

Reply:
Can you add a comment explaining `-(hit.offset + windowStart - rec.start)` for me and our future selves?